File: CPP/No198.cc
/**
 * Created by Xiaozhong on 2020/8/20.
 * Copyright (c) 2020/8/20 Xiaozhong. All rights reserved.
 */
#include "iostream"
#include "vector"

using namespace std;

class Solution {
public:
    int rob(vector<int> &nums) {
        vector<int> dp(nums.size());
        for (int i = 1; i < nums.size(); ++i) {
        }
    }
};

int main() {
    Solution s;
    vector<int> nums = {1, 2, 3, 1};
    cout << s.rob(nums) << endl;
}

class Solution_Old_Version {
public:
    int rob(vector<int> &nums) {
        if (nums.empty()) return 0;
        if (nums.size() == 1) return nums[0];
        int size = nums.size();
        vector<int> dp(size, 0);
        dp[0] = nums[0];
        dp[1] = max(nums[0], nums[1]);
        for (int i = 2; i < size; i++) dp[i] = max(dp[i - 1], dp[i - 2] + nums[i]);
        return dp[size - 1];
    }
};

File: Java/src/No198.java
/*
 * Copyright (c) 8/27/2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

public class No198 {
    public int rob(int[] nums) {
        int n = nums.length;
        int[][] dp = new int[n + 1][2];
        // 0 是指偷窃 i 位置处的屋子，1 是指不偷窃 i 位置处的屋子
        dp[1][0] = nums[0];
        dp[1][1] = 0;
        for (int i = 1; i < n; ++i) {
            dp[i + 1][0] = Math.max(dp[i - 1][0] + nums[i], dp[i][1] + nums[i]);
            dp[i + 1][1] = Math.max(dp[i][0], dp[i][1]);
        }
        return Math.max(dp[n][0], dp[n][1]);
    }

    public static void main(String[] args) {
        int[] nums = {2, 7, 9, 3, 1};
        No198 no198 = new No198();
        System.out.println(no198.rob(nums));
    }
}
